Fix doc issue #4628 (#5208)
yugabyte/yugabyte-db: docs/content/latest/api/ysql/datatypes/type_array/functions-operators/any-all.md

---
title: ANY and ALL—test if an element is in an array
linkTitle: =ANY and =ALL
headerTitle: =ANY and =ALL — test if an element is in an array
description: The =ANY and =ALL operators test if an element is in an array
menu:
  latest:
    identifier: any-all
    parent: array-functions-operators
    weight: 10
isTocNested: true
showAsideToc: true
---

## Overview

**Purpose:** Return `TRUE`, `FALSE`, or `NULL` according to the outcome of the particular comparison, `=ANY` or `=ALL`, between the input [LHS](https://en.wikipedia.org/wiki/Sides_of_an_equation) _element_ and input [RHS](https://en.wikipedia.org/wiki/Sides_of_an_equation) array. Each of the LHS and the RHS is, in general, an expression. Further, the RHS expression must be parenthesized.

> The value to which the LHS evaluates is compared, using the appropriate data type overload of the `=` operator, to each of the RHS expression's evaluated _elements_ to yield a `boolean` result (`TRUE`, `FALSE`, or `NULL`) according to the normal rules for _element_ comparison. (See the [Semantics demonstration](./#semantics-demonstration) section.)

Reviewer: ```suggestion
The value to which the LHS evaluates is compared, using the appropriate data type overload of the `=` operator, to each of the RHS expression's evaluated _elements_ to yield a `boolean` result (`TRUE`, `FALSE`, or `NULL`) according to the normal rules for _element_ comparison. (See [Semantics demonstration](./#semantics-demonstration).)
```

Author: I did a global search for `the [` and `) section and changed them all as you want. See that paragraph that used to start like this:

> The [Equality and inequality semantics](./#equality-and-inequality-semantics) section demonstrates each of the rules that this _"Comparison operators overview"_  section has stated.

It used to read well because "section" is a noun that serves well as the subject of a sentence. Now it reads badly because "Equality and inequality semantics" simply doesn't serve this purpose. But... it's what you style guide robotically insist on.

Now the doc reads like Russian where they've never heard of the definite article.

Think of `=ANY` and `=ALL` as the spellings of single [binary operators](https://en.wikipedia.org/wiki/Binary_operation). Notice that, `<> ANY`, `<> ALL`, `< ANY`, `>= ALL`, and so on are syntactically legal, and produce outcomes without error. But these outcomes are semantically meaningless.

`=ANY` is functionally equivalent to `IN` (but `IN` is illegal syntax when the RHS is an array).

`=ALL` has no functional equivalent in the way that `=ANY` is functionally equivalent to `IN`. Briefly (and leaving the possibility of `NULL` to the [semantics of the `=ALL` operator](./#the-semantics-of-the-all-operator) section, `=ALL` produces `TRUE` when every _element_ in the RHS array is equal to every other of its _elements_ and when the LHS _element_ is also equal to that value. 

> ## The semantics of the =ANY operator

Reviewer: Does this generate correctly? Assuming so, but any special characters in headers can cause unexpected results in Hugo at times.

Author: Yes, they generate properly. See, for example, [HERE](http://localhost:1313/latest/api/ysql/datatypes/type_array/functions-operators/comparison/#examples-for-each-operator). It's been out there in the Internet for ages. It looks fine. Marking as "resolved".

- If the succession of comparisons of the LHS _element_ to the RHS array's _elements_ yields no `NULL` results and yields at least one `FALSE` result, then the `=ALL` result is `FALSE`.
- If the succession of comparisons of the LHS _element_ to the RHS array's _elements_ yields at least one `NULL` result, then the `=ALL` result is `NULL`.
- In summary, _only_ if every one of the RHS array's _elements_ is `NOT NULL`, all of these _elements_ are equal to each other, and the LHS _element_ is also equal to that value, does the `=ALL` comparison yield `TRUE`.

Here is an example. It shows that you can test whether an array's _elements_ are all the same as each other without needing to know their common value or anything about the array's dimensionality.

```plpgsql
do $body$
declare
  arr constant int[] not null := '
    [2:3][4:6][7:10]={
      {
        {42,42,42,42},{42,42,42,42},{42,42,42,42}
      },
      {
        {42,42,42,42},{42,42,42,42},{42,42,42,42}
      }
    }'::int[];

   val constant int not null := (
    select unnest(arr) limit 1);

  b99 boolean not null := val =ALL (arr);
begin
  assert b99, 'assert failed';
end;
$body$;
```

> See the section [Multidimensional array of `int` values](../../literals/array-of-primitive-values/#multidimensional-array-of-int-values) for the explanation of the general syntax for the literal for a multidimensional array that specifies the lower and upper index bounds along each dimension.

Reviewer: ```suggestion
For the explanation of the general syntax for the literal for a multidimensional array that specifies the lower and upper index bounds along each dimension, see [Multidimensional array of `int` values](../../literals/array-of-primitive-values/#multidimensional-array-of-int-values).
```

Reviewer: Tell them what they're going to see before you give them a link to it. I think you know this but had a lapse. ;-)

Author: Done.

> See the section [Multidimensional array_agg() and unnest() — first overloads](..//array-agg-unnest/#multidimensional-array-agg-and-unnest-first-overloads) for the specification of the behavior `unnest()` when its actual argument is a multidimensional array.

Reviewer: ```suggestion
For the specification of the behavior `unnest()` when its actual argument is a multidimensional array, see [Multidimensional array_agg() and unnest() — first overloads](..//array-agg-unnest/#multidimensional-array-agg-and-unnest-first-overloads).
```

Author: Done.

> You might want to include this code directly in the code example presented in the section [Using an array of atomic elements](./#using-an-array-of-atomic-elements) below. It is presented stand-alone here because its design needs a specific explanation.

Reviewer: ```suggestion
You might want to include this code directly in the code example presented in [Using an array of atomic elements](./#using-an-array-of-atomic-elements) below. It is presented stand-alone here because its design needs a specific explanation.
```

## Semantics demonstration

> The first demonstration uses an array of _atomic elements_. The second demonstration uses an array of _composite elements_. And the third demonstration uses an array of _elements_ that are values of a `DOMAIN`.

Reviewer: ```suggestion
The following semantics demonstrations show the use of an array of _atomic elements_,  an array of _composite elements_, and an array of _elements_ that are values of a `DOMAIN`.
```

Reviewer: If each demonstration is long, you could add internal links to each of these. Or list them as bullets with linked titles.

Author: Done

```plpgsql
drop type if exists rt;
create type rt as (a int, b text);

do $body$
declare
  v1 constant rt := (0, 1);
  v2 constant rt := (2, 3);
  v3 constant rt := (4, 5);
  v4 constant rt := (6, 7);
  v5 constant rt := (8, 9);

  arr1 constant rt[] := array[v1, v1, v1, v1];
  arr2 constant rt[] := array[v1, v2, v3, v4];
  arr3 constant rt[] := array[v1, v1, v1, null::rt];

  b1 constant boolean not null :=     v1 =all (arr1);
  b2 constant boolean not null := not v1 =all (arr2);
  b3 constant boolean not null :=     v1 =any (arr2);
  b4 constant boolean not null := not v5 =any (arr2);

  b5 constant boolean not null :=     v1 =any (arr3);
  b6 constant boolean not null :=    (v5 =any (arr3)) is null;
  b7 constant boolean not null :=    (v1 =all (arr3)) is null;
begin
  assert
    (b1 and b2 and b3 and b4 and b5 and b6 and b7),
  'assert failed';
end;
$body$;
```

### Using an array of elements that are values of a DOMAIN

First, consider these two examples:

```plpgsql
drop domain if exists d1_t;
create domain d1_t as int
default 42 constraint d1_t_chk check(value >= 17);

drop domain if exists d2_t;
create domain d2_t as int[];
```

> It's clear that the values of _"d1_t"_, as a specialized kind of `int`, are _elements_. But what about the values of _"d2_t"_ which is a specialized kind of _array_? Critically, but somewhat counter-intuitively,  _"d2_t"_ does _not_ qualify as `anyarray`. Rather, it qualifies as `anyelement`. This code example underlines the point:

Reviewer: ```suggestion
It's clear that the values of _"d1_t"_, as a specialized kind of `int`, are _elements_. But what about the values of _"d2_t"_ which is a specialized kind of _array_? Critically, but somewhat counter-intuitively,  _"d2_t"_ does _not_ qualify as `anyarray`. Rather, it qualifies as `anyelement`. This code example emphasizes the point:
```
